core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -250,6 +252,15 @@ public String getAuthority() {
     return method;
   }
 
+  @Override
+  public SecurityLevel getSecurityLevel() {
```

Comment:
This implementation can be moved up to `ServerCall` except the call to `super.getSecurityLevel()` on line 258. Note that `getAttributes()` is defined in `ServerCall`. So your `ServerCall.getSecurityLevel()`  implementation can be

```
  Attributes attributes = getAttributes();
  if (attributes == null) {
      return SecurityLevel.NONE;
  }
  SecurityLevel securityLevel = attributes.get(ATTR_SECURITY_LEVEL);
  return securityLevel == null ? SecurityLevel.NONE : securityLevel;
```

One less file to be changed. Any issues with this implementation?

Reply:
`ATTR_SECURITY_LEVEL` is a static field and can't be imported into the abstract class `ServerCall`. Not really familiar yet with the `GrpcAttributes`, what is your approach for it?

Reply:
Hmmm, you are right. `ATTR_SECURITY_LEVEL` is from `GrpcAttributes` which is `Internal` so cannot be used from the `grpc-api`. So let's keep the structure as is.